src/platform/intel/cavs/lib/pm_memory.c

```diff
@@ -0,0 +1,202 @@
+			| operation_data->change_mask0 :
+			operation_data->ebb_current_mask0
+			& ~operation_data->change_mask0;
+		operation_data->ebb_new_mask1 =
+			operation_data->ebb_current_mask1;
+	}
+}
+
+/**
+ * \brief Write calculated masks to HW registers
+ * \param[in] ebb_data All masks data with masks to write.
+ */
+static void write_new_masks_and_check_status(struct ebb_data *operation_data)
+{
+	uint32_t status, delay_count = 256;
+	/* HSPGCTL, HSRMCTL use reverse logic - 0 means EBB is power gated */
+	io_reg_write(HSPGCTL0,  ~operation_data->ebb_new_mask0);
+	io_reg_write(HSRMCTL0,  ~operation_data->ebb_new_mask0);
+	io_reg_write(HSPGCTL1,  ~operation_data->ebb_new_mask1);
+	io_reg_write(HSRMCTL1,  ~operation_data->ebb_new_mask1);
+
+	/* Query the power status of first part of HP memory
+	 * to check whether it has been powered up. A few
+	 * cycles are needed for it to be powered up
+	 */
+	status = io_reg_read(HSPGISTS0);
+	while (status != ~operation_data->ebb_new_mask0) {
+		idelay(delay_count);
+		status = io_reg_read(HSPGISTS0);
+	}
+	/* Query the power status of second part of HP memory */
+
+	status = io_reg_read(HSPGISTS1);
+	while (status != ~operation_data->ebb_new_mask1) {
+		idelay(delay_count);
+		status = io_reg_read(HSPGISTS1);
+	}
+	/* add some delay before touch power register */
+	idelay(delay_count);
+}
+
+/**
+ * \brief Set memory power gating hw bit mask for memory banks
+ * \param[in] start_bank_id Id of first bank to be managed (inclusive) 0 based.
+ * \param[in] ending_bank_id Id of last bank to be managed (inclusive) 0 based.
+ * \param[in] bit Boolean deciding banks desired state (1 powered 0 gated).
+ */
+static void set_banks_gating(uint32_t start_bank_id, uint32_t ending_bank_id,
+			     uint32_t bit)
+{
+	struct ebb_data operation_data = {0};
+
+	operation_data.start_bank_id = start_bank_id;
+	operation_data.ending_bank_id = ending_bank_id;
+	operation_data.start_bank_id_high = start_bank_id - EBB_SEGMENT_SIZE;
+	operation_data.ending_bank_id_high = ending_bank_id - EBB_SEGMENT_SIZE;
+
+	/* Take note that if there are more banks than EBB_SEGMENT_SIZE
+	 * then those over have to be controlled by second mask
+	 * hence start_bank_id and start_bank_id_high
+	 * for calculating the banks in second segment (currently two
+	 * segments is all HW supports)
+	 */
+	shim_write(SHIM_LDOCTL, SHIM_LDOCTL_HPSRAM_LDO_ON);
+
+	set_bank_masks(&operation_data);
+
+	calculate_new_masks(&operation_data, bit);
+
+	write_new_masks_and_check_status(&operation_data);
+
+	shim_write(SHIM_LDOCTL, SHIM_LDOCTL_HPSRAM_LDO_BYPASS);
+}
+
+void set_power_gate_for_memory_address_range(void *ptr,
+					     uint32_t size, uint32_t bit)
+{
+	uint32_t start_bank_id;
+	uint32_t ending_bank_id;
+	void *end_ptr = ptr + size;
+
+	/* if ptr is not aligned to bank size change it to
+	 * closest possible memory address at the start of bank
+	 * or end for end address
+	 */
+	if ((uintptr_t)ptr % SRAM_BANK_SIZE)
+		ptr = ptr + (uintptr_t)(SRAM_BANK_SIZE
```

Comment:
What about using macros ALIGN_UP, ALIGN_DOWN introduced in Tomek PR?

Reply:
Align up performs additional checks that are not necessary/if implemented(replace whole if statement) would make code inconsistent.

Reply:
Not sure why using common macros are making code inconsistent? When if replaced = no unnecessary checks.
It will be much easier to read and we should use them as this is one of the rules from the greatest kernel coding style 😎 😎 😎

Reply:
I do not see it as easier to read but more difficult. Also which exact coding rule/style from "greatest kernel" are we talking about here?

Reply:
@dabekjakub it's not about the coding rule/style of the kernel. It's about using common sense and using simple macros with meaningful names instead of open coding them.

Reply:
Common sense is creating readable code, commenting it when it is needed. Not overusing macros and macroing anything that looks even remotely possible to do so. 
I will ask again which rule of "greatest coding style" or common sense did I break here?